File: sbe-tool/src/main/java/uk/co/real_logic/sbe/otf/OtfMessageDecoder.java
/*
 * Copyright 2014 - 2015 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.otf;

import uk.co.real_logic.agrona.DirectBuffer;
import uk.co.real_logic.sbe.ir.Token;

import java.util.List;

import static uk.co.real_logic.sbe.ir.Signal.BEGIN_FIELD;
import static uk.co.real_logic.sbe.ir.Signal.BEGIN_GROUP;
import static uk.co.real_logic.sbe.ir.Signal.BEGIN_VAR_DATA;

/**
 * On-the-fly decoder that dynamically decodes messages based on the IR for a schema.
 *
 * The contents of the messages are structurally decomposed and passed to a {@link TokenListener} for decoding the
 * primitive values.
 *
 * The design keeps all state on the stack to maximise performance and avoid object allocation. The message decoder can
 * be reused repeatably by calling {@link OtfMessageDecoder#decode(DirectBuffer, int, int, int, List, TokenListener)}
 * which is thread safe to be used across multiple threads.
 */
public class OtfMessageDecoder
{
    /**
     * Decode a message from the provided buffer based on the message schema described with IR {@link Token}s.
     *
     * @param buffer        containing the encoded message.
     * @param bufferIdx     at which the message encoding starts in the buffer.
     * @param actingVersion of the encoded message for dealing with extension fields.
     * @param blockLength   of the root message fields.
     * @param msgTokens     in IR format describing the message structure.
     * @param listener      to callback for decoding the primitive values as discovered in the structure.
     * @return the index in the underlying buffer after decoding.
     */
    public static int decode(
        final DirectBuffer buffer,
        int bufferIdx,
        final int actingVersion,
        final int blockLength,
        final List<Token> msgTokens,
        final TokenListener listener)
    {
        listener.onBeginMessage(msgTokens.get(0));

        final int numTokens = msgTokens.size();
        final int tokenIdx = decodeFields(buffer, bufferIdx, actingVersion, msgTokens, 1, numTokens, listener);
        bufferIdx += blockLength;

        final long packedValues = decodeGroups(buffer, bufferIdx, actingVersion, msgTokens, tokenIdx, numTokens, listener);

        bufferIdx = decodeData(buffer, bufferIndex(packedValues), msgTokens, tokenIndex(packedValues), numTokens, listener);

        listener.onEndMessage(msgTokens.get(numTokens - 1));

        return bufferIdx;
    }

    private static int decodeFields(
        final DirectBuffer buffer,
        final int bufferIdx,
        final int actingVersion,
        final List<Token> tokens,
        int tokenIdx,
        final int numTokens,
        final TokenListener listener)
    {
        while (tokenIdx < numTokens)
        {
            final Token fieldToken = tokens.get(tokenIdx);
            if (BEGIN_FIELD != fieldToken.signal())
            {
                break;
            }

            final int nextFieldIdx = tokenIdx + fieldToken.componentTokenCount();
            tokenIdx++;

            final Token typeToken = tokens.get(tokenIdx);
            final int offset = typeToken.offset();

            switch (typeToken.signal())
            {
                case BEGIN_COMPOSITE:
                    decodeComposite(
                        fieldToken, buffer, bufferIdx + offset, tokens, tokenIdx, nextFieldIdx - 2, actingVersion, listener);
                    break;

                case BEGIN_ENUM:
                    listener.onEnum(
                        fieldToken, buffer, bufferIdx + offset, tokens, tokenIdx, nextFieldIdx - 2, actingVersion);
                    break;

                case BEGIN_SET:
                    listener.onBitSet(
                        fieldToken, buffer, bufferIdx + offset, tokens, tokenIdx, nextFieldIdx - 2, actingVersion);
                    break;

                case ENCODING:
                    listener.onEncoding(fieldToken, buffer, bufferIdx + offset, typeToken, actingVersion);
                    break;
            }

            tokenIdx = nextFieldIdx;
        }

        return tokenIdx;
    }

    private static long decodeGroups(
        final DirectBuffer buffer,
        int bufferIdx,
        final int actingVersion,
        final List<Token> tokens,
        int tokenIdx,
        final int numTokens,
        final TokenListener listener)
    {
        while (tokenIdx < numTokens)
        {
            final Token token = tokens.get(tokenIdx);
            if (BEGIN_GROUP != token.signal())
            {
                break;
            }

            final Token blockLengthToken = tokens.get(tokenIdx + 2);
            final int blockLength = Types.getInt(
                buffer,
                bufferIdx + blockLengthToken.offset(),
                blockLengthToken.encoding().primitiveType(),
                blockLengthToken.encoding().byteOrder());

            final Token numInGroupToken = tokens.get(tokenIdx + 3);
            final int numInGroup = Types.getInt(
                buffer,
                bufferIdx + numInGroupToken.offset(),
                numInGroupToken.encoding().primitiveType(),
                numInGroupToken.encoding().byteOrder());

            final Token dimensionTypeComposite = tokens.get(tokenIdx + 1);
            bufferIdx += dimensionTypeComposite.encodedLength();

            final int beginFieldsIdx = tokenIdx + dimensionTypeComposite.componentTokenCount() + 1;

            listener.onGroupHeader(token, numInGroup);

            for (int i = 0; i < numInGroup; i++)
            {
                listener.onBeginGroup(token, i, numInGroup);

                final int afterFieldsIdx = decodeFields(
                    buffer, bufferIdx, actingVersion, tokens, beginFieldsIdx, numTokens, listener);
                bufferIdx += blockLength;

                final long packedValues = decodeGroups(
                    buffer, bufferIdx, actingVersion, tokens, afterFieldsIdx, numTokens, listener);

                bufferIdx = decodeData(
                    buffer, bufferIndex(packedValues), tokens, tokenIndex(packedValues), numTokens, listener);

                listener.onEndGroup(token, i, numInGroup);
            }

            tokenIdx += token.componentTokenCount();
        }

        return pack(bufferIdx, tokenIdx);
    }

    private static void decodeComposite(
        final Token fieldToken,
        final DirectBuffer buffer,
        final int bufferIdx,
        final List<Token> tokens,
        final int tokenIdx,
        final int toIndex,
        final int actingVersion,
        final TokenListener listener)
    {
        listener.onBeginComposite(fieldToken, tokens, tokenIdx, toIndex);

        for (int i = tokenIdx + 1; i < toIndex;)
        {
            final Token typeToken = tokens.get(i);
            final int nextFieldIdx = i + typeToken.componentTokenCount();

            final int offset = typeToken.offset();

            switch (typeToken.signal())
            {
                case BEGIN_COMPOSITE:
                    decodeComposite(
                        fieldToken, buffer, bufferIdx + offset, tokens, i, nextFieldIdx - 1, actingVersion, listener);
                    break;

                case BEGIN_ENUM:
                    listener.onEnum(
                        fieldToken, buffer, bufferIdx + offset, tokens, i, nextFieldIdx - 1, actingVersion);
                    break;

                case BEGIN_SET:
                    listener.onBitSet(
                        fieldToken, buffer, bufferIdx + offset, tokens, i, nextFieldIdx - 1, actingVersion);
                    break;

                case ENCODING:
                    listener.onEncoding(typeToken, buffer, bufferIdx + offset, typeToken, actingVersion);
                    break;
            }

            i += typeToken.componentTokenCount();
        }

        listener.onEndComposite(fieldToken, tokens, tokenIdx, toIndex);
    }

    private static int decodeData(
        final DirectBuffer buffer,
        int bufferIdx,
        final List<Token> tokens,
        int tokenIdx,
        final int numTokens,
        final TokenListener listener)
    {
        while (tokenIdx < numTokens)
        {
            final Token token = tokens.get(tokenIdx);
            if (BEGIN_VAR_DATA != token.signal())
            {
                break;
            }

            final Token lengthToken = tokens.get(tokenIdx + 2);
            final int length = Types.getInt(
                buffer,
                bufferIdx + lengthToken.offset(),
                lengthToken.encoding().primitiveType(),
                lengthToken.encoding().byteOrder());

            final Token dataToken = tokens.get(tokenIdx + 3);
            bufferIdx += dataToken.offset();

            listener.onVarData(token, buffer, bufferIdx, length, dataToken);

            bufferIdx += length;
            tokenIdx += token.componentTokenCount();
        }

        return bufferIdx;
    }

    private static long pack(final int bufferIndex, final int tokenIndex)
    {
        return ((long)bufferIndex << 32) | tokenIndex;
    }

    private static int bufferIndex(final long packedValues)
    {
        return (int)(packedValues >>> 32);
    }

    private static int tokenIndex(final long packedValues)
    {
        return (int)packedValues;
    }
}

File: sbe-tool/src/main/golang/otf/otfmessagedecoder.go
package otf

import "fmt"

type OnBeginMessage func(token Token)
type OnEndMessage func(token Token)
type OnEncoding func(token Token, buffer []byte, actingVersion uint64)
type OnEnum func(token Token, buffer []byte, tokens []Token, fromIndex int, toIndex int, actingVersion uint64)
type OnBitSet func(token Token, buffer []byte, tokens []Token, fromIndex int, toIndex int, actingVersion uint64)
type OnBeginComposite func(token Token, tokens []Token, fromIndex int, toIndex int)
type OnEndComposite func(token Token, tokens []Token, fromIndex int, toIndex int)
type OnGroupHeader func(token Token, numInGroup uint64)
type OnBeginGroup func(token Token, groupIndex uint64, numInGroup uint64)
type OnEndGroup func(token Token, groupIndex uint64, numInGroup uint64)
type OnVarData func(token Token, buffer []byte, length uint64, typeToken Token)

type TokenListener interface {
	OnBeginMessage(token Token)
	OnEndMessage(token Token)
	OnEncoding(fieldToken Token, buffer []byte, typeToken Token, actingVersion uint64)
	OnEnum(fieldToken Token, buffer []byte, tokens []Token, fromIndex int, toIndex int, actingVersion uint64)
	OnBitSet(fieldToken Token, buffer []byte, tokens []Token, fromIndex int, toIndex int, actingVersion uint64)
	OnBeginComposite(fieldToken Token, tokens []Token, fromIndex int, toIndex int)
	OnEndComposite(fieldToken Token, tokens []Token, fromIndex int, toIndex int)
	OnGroupHeader(token Token, numInGroup uint64)
	OnBeginGroup(token Token, groupIndex uint64, numInGroup uint64)
	OnEndGroup(token Token, groupIndex uint64, numInGroup uint64)
	OnVarData(fieldToken Token, buffer []byte, length uint64, typeToken Token)
}

func DecodeComposite(
	fieldToken Token,
	buffer []byte,
	bufferIndex int,
	length uint64,
	tokens []Token,
	tokenIndex int,
	toIndex int,
	actingVersion uint64,
	listener TokenListener) int {
	listener.OnBeginComposite(fieldToken, tokens, tokenIndex, toIndex)
	for i := tokenIndex + 1; i < toIndex; {
		token := tokens[i]
		nextFieldIndex := i + int(token.ComponentTokenCount())
		offset := int(token.Offset())
		switch token.Signal() {
		case SignalBeginComposite:
			DecodeComposite(
				fieldToken,
				buffer,
				bufferIndex+offset,
				length,
				tokens,
				i,
				nextFieldIndex-1,
				actingVersion,
				listener)
		case SignalBeginEnum:
			listener.OnEnum(
				fieldToken,
				buffer[bufferIndex+offset:],
				tokens,
				i,
				nextFieldIndex-1,
				actingVersion)
		case SignalBeginSet:
			listener.OnBitSet(
				fieldToken,
				buffer[bufferIndex+offset:],
				tokens,
				i,
				nextFieldIndex-1,
				actingVersion)
		case SignalEncoding:
			listener.OnEncoding(token, buffer[bufferIndex+offset:], token, actingVersion)
		}
		i += int(token.ComponentTokenCount())
	}
	listener.OnEndComposite(fieldToken, tokens, tokenIndex, toIndex)
	return tokenIndex
}

func DecodeFields(
	buffer []byte,
	bufferIndex int,
	length uint64,
	actingVersion uint64,
	tokens []Token,
	tokenIndex int,
	numTokens int,
	listener TokenListener) int {

	for tokenIndex < numTokens {
		fieldToken := tokens[tokenIndex]
		if SignalBeginField != fieldToken.Signal() {
			break
		}
		nextFieldIndex := tokenIndex + int(fieldToken.ComponentTokenCount())
		tokenIndex++

		typeToken := tokens[tokenIndex]
		offset := bufferIndex + int(typeToken.Offset())
		switch typeToken.Signal() {
		case SignalBeginComposite:
			DecodeComposite(
				fieldToken,
				buffer,
				offset,
				length,
				tokens,
				tokenIndex,
				nextFieldIndex-2,
				actingVersion,
				listener)
		case SignalBeginEnum:
			listener.OnEnum(
				fieldToken,
				buffer[offset:],
				tokens,
				tokenIndex,
				nextFieldIndex-2,
				actingVersion)
		case SignalBeginSet:
			listener.OnBitSet(
				fieldToken,
				buffer[offset:],
				tokens,
				tokenIndex,
				nextFieldIndex-2,
				actingVersion)
		case SignalEncoding:
			listener.OnEncoding(fieldToken, buffer[offset:], typeToken, actingVersion)
		}
		tokenIndex = nextFieldIndex
	}

	return tokenIndex
}

func DecodeData(
	buffer []byte,
	bufferIndex int,
	length uint64,
	tokens []Token,
	tokenIndex int,
	numTokens int,
	actingVersion uint64,
	listener TokenListener) int {
	for tokenIndex < numTokens {
		token := tokens[tokenIndex]
		if SignalBeginVarData != token.Signal() {
			break
		}
		isPresent := token.TokenVersion() <= int32(actingVersion)
		lengthToken := tokens[tokenIndex+2]
		dataToken := tokens[tokenIndex+3]
		if (uint64(bufferIndex) + uint64(dataToken.Offset())) > length {
			panic("length too short for data length field")
		}
		dataLength := uint64(0)
		if isPresent {
			lengthTokenEncoding := lengthToken.Encoding()

			var err error
			dataLength, err = lengthTokenEncoding.GetAsUInt(buffer[bufferIndex+int(lengthToken.Offset()):])
			if err != nil {
				panic(fmt.Errorf("invalid length encoding %s", err))
			}
			bufferIndex += int(dataToken.Offset())
		}

		if (uint64(bufferIndex) + dataLength) > length {
			panic("length too short for data field")
		}
		listener.OnVarData(token, buffer[bufferIndex:], dataLength, dataToken)

		bufferIndex += int(dataLength)
		tokenIndex += int(token.ComponentTokenCount())
	}
	return bufferIndex
}

func DecodeGroups(
	buffer []byte,
	bufferIndex int,
	length uint64,
	actingVersion uint64,
	tokens []Token,
	tokenIndex int,
	numTokens int,
	listener TokenListener) (int, int) {
	for tokenIndex < numTokens {
		token := tokens[tokenIndex]
		if SignalBeginGroup != token.Signal() {
			break
		}
		isPresent := token.TokenVersion() <= int32(actingVersion)
		dimensionsTypeComposite := tokens[tokenIndex+1]
		dimensionsLength := int(dimensionsTypeComposite.EncodedLength())
		if (uint64(bufferIndex) + uint64(dimensionsLength)) > length {
			panic("length too short for dimensions composite")
		}

		blockLengthToken := tokens[tokenIndex+2]
		numInGroupToken := tokens[tokenIndex+3]
		blockLength := uint64(0)
		numInGroup := uint64(0)
		if isPresent {
			blockLengthTokenEncoding := blockLengthToken.Encoding()
			var err error
			blockLength, err = blockLengthTokenEncoding.GetAsUInt(buffer[bufferIndex+int(blockLengthToken.Offset()):])
			if err != nil {
				panic(fmt.Errorf("invalid block length encoding %s", err))
			}
			numInGroupTokenEncoding := numInGroupToken.Encoding()
			numInGroup, err = numInGroupTokenEncoding.GetAsUInt(buffer[bufferIndex+int(numInGroupToken.Offset()):])
			if err != nil {
				panic(fmt.Errorf("invalid num in group encoding %s", err))
			}
			bufferIndex += dimensionsLength
		}

		beginFieldsIndex := tokenIndex + int(dimensionsTypeComposite.ComponentTokenCount()) + 1
		listener.OnGroupHeader(token, numInGroup)
		for i := uint64(0); i < numInGroup; i++ {
			listener.OnBeginGroup(token, i, numInGroup)
			if (uint64(bufferIndex) + blockLength) > length {
				panic("length too short for group blockLength")
			}
			afterFieldsIndex := DecodeFields(
				buffer,
				bufferIndex,
				length,
				actingVersion,
				tokens,
				beginFieldsIndex,
				numTokens,
				listener)
			bufferIndex += int(blockLength)
			groupIndex, groupNumTokens := DecodeGroups(
				buffer,
				bufferIndex,
				length,
				actingVersion,
				tokens,
				afterFieldsIndex,
				numTokens,
				listener)
			bufferIndex = DecodeData(
				buffer,
				groupIndex,
				length,
				tokens,
				groupNumTokens,
				numTokens,
				actingVersion,
				listener,
			)
			listener.OnEndGroup(token, i, numInGroup)
		}
		tokenIndex += int(token.ComponentTokenCount())
	}
	return bufferIndex, tokenIndex
}

func Decode(
	buffer []byte,
	actingVersion uint64,
	blockLength uint64,
	tokens []Token,
	listener TokenListener) int {
	listener.OnBeginMessage(tokens[0])
	length := uint64(len(buffer))
	if length < blockLength {
		panic("length too short for message blockLength")
	}
	numTokens := len(tokens)
	tokenIndex := DecodeFields(
		buffer,
		0,
		length,
		actingVersion,
		tokens,
		1,
		numTokens,
		listener,
	)
	bufferIndex := int(blockLength)
	groupIndex, groupNumTokens := DecodeGroups(
		buffer,
		bufferIndex,
		length,
		actingVersion,
		tokens,
		tokenIndex,
		numTokens,
		listener,
	)
	bufferIndex = DecodeData(
		buffer,
		groupIndex,
		length,
		tokens,
		groupNumTokens,
		numTokens,
		actingVersion,
		listener,
	)
	listener.OnEndMessage(tokens[numTokens-1])
	return bufferIndex
}
